vpr/src/route/overuse_report.cpp

```diff
@@ -121,6 +112,80 @@ void generate_overused_nodes_to_congested_net_lookup(std::map<RRNodeId, std::set
     }
 }
 
+static void report_overused_ipin_opin(std::ostream& os, RRNodeId node_id) {
+    const auto& device_ctx = g_vpr_ctx.device();
+    const auto& place_ctx = g_vpr_ctx.placement();
+
+    auto grid_x = device_ctx.rr_nodes.node_xlow(node_id);
+    auto grid_y = device_ctx.rr_nodes.node_ylow(node_id);
+    VTR_ASSERT_MSG(
+        grid_x == device_ctx.rr_nodes.node_xhigh(node_id) && grid_y == device_ctx.rr_nodes.node_yhigh(node_id),
+        "Non-track RR node should not span across multiple grid blocks.");
+
+    os << "Pin number = " << device_ctx.rr_nodes.node_pin_num(node_id) << '\n';
+    os << "Side = " << device_ctx.rr_nodes.node_side_string(node_id) << "\n\n";
+
+    //Add block type for IPINs/OPINs in overused rr-node report
+    const auto& clb_nlist = g_vpr_ctx.clustering().clb_nlist;
+    auto& grid_info = place_ctx.grid_blocks[grid_x][grid_y];
+    auto& grid_blocks = grid_info.blocks;
+
+    os << "Grid location: X = " << grid_x << ", Y = " << grid_y << '\n';
+    os << "Number of blocks currently at this grid location = " << grid_info.usage << '\n';
+    for (size_t iblock = 0; iblock < grid_blocks.size(); ++iblock) {
+        ClusterBlockId block_id = grid_blocks[iblock];
+        os << "Block #" << iblock << ": ";
```

Comment:
I think you need to add a check for empty blocks here.  grid_blocks.size gives the capacity of that (x,y) location (how many things could be placed there).  But you could have locations where nothing was placed.  See 
https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/ed1a373d9c3409eea098612b6530f734c125ca03/vpr/src/util/vpr_utils.cpp#L192
and the code just above:
            for (int z = 0; z < device_ctx.grid[x][y].type->capacity; ++z) {
                grid_block.blocks[z] = EMPTY_BLOCK_ID;
            }
        }

So you should check for EMPTY_BLOCK_ID.  I think adding a short function (grid_block.empty(i)) to t_grid_block would be a nice way to code this, rather than doing a check for EMPTY_BLOCK_ID (that small function would localize the EMPTY_BLOCK_ID check).